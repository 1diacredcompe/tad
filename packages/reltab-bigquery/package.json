{
  "name": "reltab-bigquery",
  "version": "0.10.0",
  "description": "Google BigQuery driver for reltab",
  "main": "dist/reltab-bigquery.js",
  "types": "dist/reltab-biguery.d.ts",
  "scripts": {
    "prepublish": "npm run build",
    "clean": "rimraf dist",
    "build": "tsc",
    "test": "jest --config jest.config.json"
  },
  "keywords": [
    "relational",
    "sql",
    "database",
    "bigquery"
  ],
  "author": "Antony Courtney <antony@antonycourtney.com>",
  "license": "MIT",
  "dependencies": {
    "loglevel": "^1.6.8",
    "reltab": "^0.10.0",
    "@google-cloud/bigquery": "4.7.0"
  },
  "devDependencies": {
    "@types/jest": "^25.2.1",
    "@types/node": "^13.11.0",
    "jest": "^25.2.7",
    "mkdirp": "^1.0.4",
    "prettier": "^2.0.2",
    "ts-jest": "^25.3.1",
    "typescript": "^3.8.3"
  }
}
