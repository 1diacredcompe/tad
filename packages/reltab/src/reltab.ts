export * from "./dialectRegistry";
export * from "./Connection";
export * from "./defs";
export * from "./FilterExp";
export * from "./QueryExp";
export * from "./Schema";
export * from "./TableRep";
export * from "./AggFn";
export * from "./ColumnType";
export * from "./DataSource";
