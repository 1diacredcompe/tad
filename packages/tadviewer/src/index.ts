/*
export * from "./components/AggPanel";
export * from "./components/ColumnList";
export * from "./components/ColumnRow";
export * from "./components/ColumnSelector";
export * from "./components/DisplayOrderPanel";
export * from "./components/FilterEditor";
export * from "./components/FilterEditorRow";
export * from "./components/Footer";
export * from "./components/FormatPanel";
export * from "./components/GridPane";
export * from "./components/IndeterminateCheckbox";
export * from "./components/LoadingModal";
export * from "./components/NumFormatPanel";
export * from "./components/PivotOrderPanel";
export * from "./components/Sidebar";
export * from "./components/SingleColumnSelect";
export * from "./components/SortOrderPanel";
export * from "./components/TextFormatPanel";
*/

export * from "./AppState";
export * from "./components/AppPane";
