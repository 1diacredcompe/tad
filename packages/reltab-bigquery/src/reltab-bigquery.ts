import * as log from "loglevel";
import { TableRep, QueryExp, Schema, tableQuery } from "reltab";
import {
  TableInfoMap,
  TableInfo,
  ValExp,
  Row,
  AggColSpec,
  SubExp,
  ColumnMetaMap,
  ColumnMapInfo,
  Connection,
  BigQueryDialect,
} from "reltab";
import { BigQuery, Dataset } from "@google-cloud/bigquery";

const LOCATION = "US";

function assertDefined<A>(x: A | undefined | null): A {
  if (x == null) {
    throw new Error("unexpected null value");
  }

  return x;
}

export interface BigQueryConnectionOptions {
  showQueries?: boolean;
}
export class BigQueryConnection implements Connection {
  projectId: string;
  datasetName: string;
  bigquery: BigQuery;
  bigquery_meta: BigQuery;
  dataset: Dataset;
  tableMap: TableInfoMap;
  showQueries: boolean;

  constructor(
    projectId: string,
    datasetName: string,
    options?: BigQueryConnectionOptions
  ) {
    this.projectId = projectId;
    this.datasetName = datasetName;

    this.bigquery = new BigQuery();
    this.bigquery_meta = new BigQuery({ projectId, location: LOCATION });
    this.dataset = this.bigquery_meta.dataset(datasetName);
    this.tableMap = {};
    this.showQueries =
      options != null && options.showQueries != null
        ? options.showQueries
        : false;
  }

  async evalQuery(
    query: QueryExp,
    offset: number = -1,
    limit: number = -1
  ): Promise<TableRep> {
    let t0 = process.hrtime();
    const schema = query.getSchema(this.tableMap);
    const sqlQuery = query.toSql(
      BigQueryDialect.getInstance(),
      this.tableMap,
      offset,
      limit
    );
    let t1 = process.hrtime(t0);
    const [t1s, t1ns] = t1;

    if (this.showQueries) {
      log.info("time to generate sql: %ds %dms", t1s, t1ns / 1e6);
      log.debug("SqliteContext.evalQuery: evaluating:");
      log.debug(sqlQuery);
    }

    const t2 = process.hrtime();
    const [dbRows] = await this.bigquery.query({
      query: sqlQuery,
      location: LOCATION,
    });
    const rows = dbRows as Row[];
    const t3 = process.hrtime(t2);
    const [t3s, t3ns] = t3;
    const t4pre = process.hrtime();
    const ret = new TableRep(schema, rows);
    const t4 = process.hrtime(t4pre);
    const [t4s, t4ns] = t4;

    if (this.showQueries) {
      log.info("time to run query: %ds %dms", t3s, t3ns / 1e6);
      log.info("time to mk table rep: %ds %dms", t4s, t4ns / 1e6);
    }

    return ret;
  }

  async rowCount(query: QueryExp): Promise<number> {
    let t0 = process.hrtime();
    const countSql = query.toCountSql(
      BigQueryDialect.getInstance(),
      this.tableMap
    );
    let t1 = process.hrtime(t0);
    const [t1s, t1ns] = t1;

    if (this.showQueries) {
      log.info("time to generate sql: %ds %dms", t1s, t1ns / 1e6);
      log.debug("SqliteContext.evalQuery: evaluating:");
      log.debug(countSql);
    }

    const t2 = process.hrtime();
    const [dbRows] = await this.bigquery.query({
      query: countSql,
      location: LOCATION,
    });
    const t3 = process.hrtime(t2);
    const [t3s, t3ns] = t3;
    log.info("time to run query: %ds %dms", t3s, t3ns / 1e6);
    const ret = Number.parseInt(dbRows[0].rowCount);
    return ret;
  }

  async dbGetTableInfo(
    projectId: string,
    datasetName: string,
    baseTableName: string
  ): Promise<TableInfo> {
    let dataset: Dataset;
    if (projectId === this.projectId && datasetName === this.datasetName) {
      dataset = this.dataset;
    } else {
      const bigquery = new BigQuery({ projectId, location: LOCATION });
      dataset = bigquery.dataset(datasetName);
    }
    const table = dataset.table(baseTableName);
    const [metadata, apiResponse] = await table.getMetadata();
    console.log("metadata: ", JSON.stringify(metadata, null, 2));
    const fields = metadata.schema.fields;
    const columnIds: string[] = fields.map((f: any) => f.name);

    const extendCMap = (
      cmm: ColumnMetaMap,
      field: any,
      idx: number
    ): ColumnMetaMap => {
      const cnm = field.name;
      const cType = field.type.toLocaleLowerCase();

      if (cType == null) {
        log.error(
          'mkTableInfo: No column type for "' + cnm + '", index: ' + idx
        );
      }

      const cmd = {
        displayName: cnm,
        type: assertDefined(cType),
      };
      cmm[cnm] = cmd;
      return cmm;
    };

    const cmMap = fields.reduce(extendCMap, {});
    const schema = new Schema(columnIds, cmMap);
    return {
      tableName: projectId + "." + datasetName + "." + baseTableName,
      schema,
    };
  }

  async getTableInfo(tableName: string): Promise<TableInfo> {
    let ti = this.tableMap[tableName];
    if (!ti) {
      const [projectId, datasetName, baseTableName] = tableName.split(".");
      ti = await this.dbGetTableInfo(projectId, datasetName, baseTableName);
      if (ti) {
        this.tableMap[tableName] = ti;
      }
    }
    return ti;
  }
}
