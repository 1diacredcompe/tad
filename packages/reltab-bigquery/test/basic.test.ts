import * as reltab from "reltab";
import { BigQueryConnection } from "../src/reltab-bigquery";
import * as log from "loglevel";

const PROJECT_ID = "";

beforeAll(() => {
  log.setLevel(log.levels.DEBUG);
});

test("t0 - basic functionality", async () => {
  const rtc = new BigQueryConnection(
    "bigquery-public-data",
    "covid19_jhu_csse",
    { showQueries: true }
  );

  const ti = await rtc.getTableInfo(
    "bigquery-public-data.covid19_jhu_csse.summary"
  );
  console.log("tableInfo: ", ti);

  console.log("basic table query:");
  const q1 = reltab.tableQuery("bigquery-public-data.covid19_jhu_csse.summary");
  const q1res = await rtc.evalQuery(q1, 0, 100);
  console.log("q1 query result: ", q1res);
});
