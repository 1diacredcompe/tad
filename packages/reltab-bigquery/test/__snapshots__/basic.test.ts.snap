// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic groupBy 1`] = `
TableRep {
  "rowData": Array [
    Object {
      "JobFamily": "Police",
      "TCOE": 127660,
      "Title": "Revenue Protection Guard",
    },
    Object {
      "JobFamily": "Safety",
      "TCOE": 211941,
      "Title": "Chief Safety Officer",
    },
    Object {
      "JobFamily": "Legal & Paralegal",
      "TCOE": 321543,
      "Title": "General Counsel",
    },
    Object {
      "JobFamily": "Legal & Paralegal",
      "TCOE": 445028,
      "Title": "Senior Attorney",
    },
    Object {
      "JobFamily": "Executive Management",
      "TCOE": 819582,
      "Title": "General Manager",
    },
    Object {
      "JobFamily": "Executive Management",
      "TCOE": 288160,
      "Title": "Assistant General Manager Operations",
    },
    Object {
      "JobFamily": "Executive Management",
      "TCOE": 281548,
      "Title": "Deputy General Manager",
    },
    Object {
      "JobFamily": "Finance & Accounting",
      "TCOE": 287139,
      "Title": "Controller-Treasurer",
    },
    Object {
      "JobFamily": "Transportation Operations",
      "TCOE": 238211,
      "Title": "Chief Transportation Officer",
    },
    Object {
      "JobFamily": "Transportation Operations",
      "TCOE": 118799,
      "Title": "Train Operator",
    },
    Object {
      "JobFamily": "Transportation Operations",
      "TCOE": 352988,
      "Title": "Station Agent",
    },
    Object {
      "JobFamily": "Public Affairs & Marketing",
      "TCOE": 181955,
      "Title": "Department Manager Gov't & Comm Rel",
    },
    Object {
      "JobFamily": "Engineering & Systems Engineering",
      "TCOE": 209709,
      "Title": "Group Manager, Capital Projects",
    },
    Object {
      "JobFamily": "Engineering & Systems Engineering",
      "TCOE": 212535,
      "Title": "Group Manager Systems Capital Prog",
    },
    Object {
      "JobFamily": "Maintenance, Vehicle & Facilities",
      "TCOE": 128332,
      "Title": "Wayside Inspector",
    },
    Object {
      "JobFamily": "Maintenance, Vehicle & Facilities",
      "TCOE": 124962,
      "Title": "Structures Inspector",
    },
    Object {
      "JobFamily": "Maintenance, Vehicle & Facilities",
      "TCOE": 123432,
      "Title": "Vehicle Inspector",
    },
    Object {
      "JobFamily": "Maintenance, Vehicle & Facilities",
      "TCOE": 104260,
      "Title": "Transit Vehicle Mechanic",
    },
    Object {
      "JobFamily": "Maintenance, Vehicle & Facilities",
      "TCOE": 113775,
      "Title": "Electrician",
    },
  ],
  "schema": Schema {
    "_sortedColumns": null,
    "columnIndices": Object {
      "JobFamily": 0,
      "TCOE": 2,
      "Title": 1,
    },
    "columnMetadata": Object {
      "Base": Object {
        "displayName": "Base",
        "type": "integer",
      },
      "JobFamily": Object {
        "displayName": "JobFamily",
        "type": "string",
      },
      "Name": Object {
        "displayName": "Name",
        "type": "string",
      },
      "TCOE": Object {
        "displayName": "TCOE",
        "type": "integer",
      },
      "Title": Object {
        "displayName": "Title",
        "type": "string",
      },
      "Union": Object {
        "displayName": "Union",
        "type": "string",
      },
    },
    "columns": Array [
      "JobFamily",
      "Title",
      "TCOE",
    ],
  },
}
`;

exports[`basic project operator 1`] = `
TableRep {
  "rowData": Array [
    Object {
      "Base": 75701,
      "JobFamily": "Police",
      "Name": "Ornellas, Robert",
      "TCOE": 127660,
      "Title": "Revenue Protection Guard",
      "Union": "BPOA",
    },
    Object {
      "Base": 152125,
      "JobFamily": "Safety",
      "Name": "Hardy, Leonard",
      "TCOE": 211941,
      "Title": "Chief Safety Officer",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 238269,
      "JobFamily": "Legal & Paralegal",
      "Name": "Burrows, Matthew",
      "TCOE": 321543,
      "Title": "General Counsel",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 165470,
      "JobFamily": "Legal & Paralegal",
      "Name": "Lee, Thomas",
      "TCOE": 224345,
      "Title": "Senior Attorney",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 160289,
      "JobFamily": "Legal & Paralegal",
      "Name": "Nuetzel, Victoria",
      "TCOE": 220683,
      "Title": "Senior Attorney",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 312461,
      "JobFamily": "Executive Management",
      "Name": "Crunican, Grace",
      "TCOE": 399921,
      "Title": "General Manager",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 298700,
      "JobFamily": "Executive Management",
      "Name": "Dugger, Dorothy",
      "TCOE": 419661,
      "Title": "General Manager",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 212647,
      "JobFamily": "Executive Management",
      "Name": "Oversier, Paul",
      "TCOE": 288160,
      "Title": "Assistant General Manager Operations",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 206914,
      "JobFamily": "Executive Management",
      "Name": "DeVaughn, Marcia",
      "TCOE": 281548,
      "Title": "Deputy General Manager",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 208744,
      "JobFamily": "Finance & Accounting",
      "Name": "Schroeder, Scott",
      "TCOE": 287139,
      "Title": "Controller-Treasurer",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 170383,
      "JobFamily": "Transportation Operations",
      "Name": "Crespo, Rodolfo",
      "TCOE": 238211,
      "Title": "Chief Transportation Officer",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 61184,
      "JobFamily": "Transportation Operations",
      "Name": "Smith, Westley",
      "TCOE": 118799,
      "Title": "Train Operator",
      "Union": "ATU",
    },
    Object {
      "Base": 60790,
      "JobFamily": "Transportation Operations",
      "Name": "Serrianni, Cheryl",
      "TCOE": 132030,
      "Title": "Station Agent",
      "Union": "ATU",
    },
    Object {
      "Base": 60781,
      "JobFamily": "Transportation Operations",
      "Name": "Nagel, Kevin",
      "TCOE": 105854,
      "Title": "Station Agent",
      "Union": "ATU",
    },
    Object {
      "Base": 60771,
      "JobFamily": "Transportation Operations",
      "Name": "Riggs, Joshua",
      "TCOE": 115104,
      "Title": "Station Agent",
      "Union": "ATU",
    },
    Object {
      "Base": 130035,
      "JobFamily": "Public Affairs & Marketing",
      "Name": "Hamill, Kerry",
      "TCOE": 181955,
      "Title": "Department Manager Gov't & Comm Rel",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 157300,
      "JobFamily": "Engineering & Systems Engineering",
      "Name": "Ruffa, Frank",
      "TCOE": 209709,
      "Title": "Group Manager, Capital Projects",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 152434,
      "JobFamily": "Engineering & Systems Engineering",
      "Name": "Dupont, Jean-Luc",
      "TCOE": 212535,
      "Title": "Group Manager Systems Capital Prog",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 77446,
      "JobFamily": "Maintenance, Vehicle & Facilities",
      "Name": "Hui, Sam",
      "TCOE": 128332,
      "Title": "Wayside Inspector",
      "Union": "SEIU",
    },
    Object {
      "Base": 77446,
      "JobFamily": "Maintenance, Vehicle & Facilities",
      "Name": "Reinhardt, Michael",
      "TCOE": 124962,
      "Title": "Structures Inspector",
      "Union": "SEIU",
    },
    Object {
      "Base": 77446,
      "JobFamily": "Maintenance, Vehicle & Facilities",
      "Name": "Estrada, Phillip",
      "TCOE": 123432,
      "Title": "Vehicle Inspector",
      "Union": "SEIU",
    },
    Object {
      "Base": 74068,
      "JobFamily": "Maintenance, Vehicle & Facilities",
      "Name": "Smith, Daniel",
      "TCOE": 104260,
      "Title": "Transit Vehicle Mechanic",
      "Union": "SEIU",
    },
    Object {
      "Base": 59817,
      "JobFamily": "Maintenance, Vehicle & Facilities",
      "Name": "Knight, Joshua",
      "TCOE": 113775,
      "Title": "Electrician",
      "Union": "SEIU",
    },
  ],
  "schema": Schema {
    "_sortedColumns": null,
    "columnIndices": Object {
      "Base": 4,
      "JobFamily": 0,
      "Name": 3,
      "TCOE": 5,
      "Title": 1,
      "Union": 2,
    },
    "columnMetadata": Object {
      "Base": Object {
        "displayName": "Base",
        "type": "integer",
      },
      "JobFamily": Object {
        "displayName": "JobFamily",
        "type": "string",
      },
      "Name": Object {
        "displayName": "Name",
        "type": "string",
      },
      "TCOE": Object {
        "displayName": "TCOE",
        "type": "integer",
      },
      "Title": Object {
        "displayName": "Title",
        "type": "string",
      },
      "Union": Object {
        "displayName": "Union",
        "type": "string",
      },
    },
    "columns": Array [
      "JobFamily",
      "Title",
      "Union",
      "Name",
      "Base",
      "TCOE",
    ],
  },
}
`;

exports[`extended groupBy 1`] = `
TableRep {
  "rowData": Array [
    Object {
      "TCOE": 127660,
      "Title": "Revenue Protection Guard",
      "_pivot": "Police",
    },
    Object {
      "TCOE": 211941,
      "Title": "Chief Safety Officer",
      "_pivot": "Safety",
    },
    Object {
      "TCOE": 321543,
      "Title": "General Counsel",
      "_pivot": "Legal & Paralegal",
    },
    Object {
      "TCOE": 445028,
      "Title": "Senior Attorney",
      "_pivot": "Legal & Paralegal",
    },
    Object {
      "TCOE": 819582,
      "Title": "General Manager",
      "_pivot": "Executive Management",
    },
    Object {
      "TCOE": 288160,
      "Title": "Assistant General Manager Operations",
      "_pivot": "Executive Management",
    },
    Object {
      "TCOE": 281548,
      "Title": "Deputy General Manager",
      "_pivot": "Executive Management",
    },
    Object {
      "TCOE": 287139,
      "Title": "Controller-Treasurer",
      "_pivot": "Finance & Accounting",
    },
    Object {
      "TCOE": 238211,
      "Title": "Chief Transportation Officer",
      "_pivot": "Transportation Operations",
    },
    Object {
      "TCOE": 118799,
      "Title": "Train Operator",
      "_pivot": "Transportation Operations",
    },
    Object {
      "TCOE": 352988,
      "Title": "Station Agent",
      "_pivot": "Transportation Operations",
    },
    Object {
      "TCOE": 181955,
      "Title": "Department Manager Gov't & Comm Rel",
      "_pivot": "Public Affairs & Marketing",
    },
    Object {
      "TCOE": 209709,
      "Title": "Group Manager, Capital Projects",
      "_pivot": "Engineering & Systems Engineering",
    },
    Object {
      "TCOE": 212535,
      "Title": "Group Manager Systems Capital Prog",
      "_pivot": "Engineering & Systems Engineering",
    },
    Object {
      "TCOE": 128332,
      "Title": "Wayside Inspector",
      "_pivot": "Maintenance, Vehicle & Facilities",
    },
    Object {
      "TCOE": 124962,
      "Title": "Structures Inspector",
      "_pivot": "Maintenance, Vehicle & Facilities",
    },
    Object {
      "TCOE": 123432,
      "Title": "Vehicle Inspector",
      "_pivot": "Maintenance, Vehicle & Facilities",
    },
    Object {
      "TCOE": 104260,
      "Title": "Transit Vehicle Mechanic",
      "_pivot": "Maintenance, Vehicle & Facilities",
    },
    Object {
      "TCOE": 113775,
      "Title": "Electrician",
      "_pivot": "Maintenance, Vehicle & Facilities",
    },
  ],
  "schema": Schema {
    "_sortedColumns": null,
    "columnIndices": Object {
      "TCOE": 2,
      "Title": 1,
      "_pivot": 0,
    },
    "columnMetadata": Object {
      "TCOE": Object {
        "displayName": "TCOE",
        "type": "integer",
      },
      "Title": Object {
        "displayName": "Title",
        "type": "string",
      },
      "_pivot": Object {
        "displayName": "JobFamily",
        "type": "string",
      },
    },
    "columns": Array [
      "_pivot",
      "Title",
      "TCOE",
    ],
  },
}
`;

exports[`initial aggtree Test 1`] = `
TableRep {
  "rowData": Array [
    Object {
      "Base": 611161,
      "JobFamily": "Executive Management",
      "Name": null,
      "Rec": 2,
      "TCOE": 819582,
      "Title": "General Manager",
      "Union": "Non-Represented",
      "_depth": 2,
      "_isRoot": 0,
      "_path0": "Executive Management",
      "_path1": "General Manager",
      "_pivot": "General Manager",
      "_sortVal_0": 1,
      "_sortVal_1": 1,
      "_sortVal_2": 0,
    },
    Object {
      "Base": 212647,
      "JobFamily": "Executive Management",
      "Name": "Oversier, Paul",
      "Rec": 1,
      "TCOE": 288160,
      "Title": "Assistant General Manager Operations",
      "Union": "Non-Represented",
      "_depth": 2,
      "_isRoot": 0,
      "_path0": "Executive Management",
      "_path1": "Assistant General Manager Operations",
      "_pivot": "Assistant General Manager Operations",
      "_sortVal_0": 1,
      "_sortVal_1": 1,
      "_sortVal_2": 0,
    },
    Object {
      "Base": 206914,
      "JobFamily": "Executive Management",
      "Name": "DeVaughn, Marcia",
      "Rec": 1,
      "TCOE": 281548,
      "Title": "Deputy General Manager",
      "Union": "Non-Represented",
      "_depth": 2,
      "_isRoot": 0,
      "_path0": "Executive Management",
      "_path1": "Deputy General Manager",
      "_pivot": "Deputy General Manager",
      "_sortVal_0": 1,
      "_sortVal_1": 1,
      "_sortVal_2": 0,
    },
  ],
  "schema": Schema {
    "_sortedColumns": null,
    "columnIndices": Object {
      "Base": 4,
      "JobFamily": 0,
      "Name": 3,
      "Rec": 6,
      "TCOE": 5,
      "Title": 1,
      "Union": 2,
      "_depth": 7,
      "_isRoot": 9,
      "_path0": 13,
      "_path1": 14,
      "_pivot": 8,
      "_sortVal_0": 10,
      "_sortVal_1": 11,
      "_sortVal_2": 12,
    },
    "columnMetadata": Object {
      "Base": Object {
        "displayName": "Base",
        "type": "integer",
      },
      "JobFamily": Object {
        "displayName": "JobFamily",
        "type": "string",
      },
      "Name": Object {
        "displayName": "Name",
        "type": "string",
      },
      "Rec": Object {
        "type": "integer",
      },
      "TCOE": Object {
        "displayName": "TCOE",
        "type": "integer",
      },
      "Title": Object {
        "displayName": "Title",
        "type": "string",
      },
      "Union": Object {
        "displayName": "Union",
        "type": "string",
      },
      "_depth": Object {
        "type": "integer",
      },
      "_isRoot": Object {
        "type": "boolean",
      },
      "_path0": Object {
        "type": "text",
      },
      "_path1": Object {
        "type": "text",
      },
      "_pivot": Object {
        "displayName": "_pivot",
        "type": "text",
      },
      "_sortVal_0": Object {
        "type": "integer",
      },
      "_sortVal_1": Object {
        "type": "integer",
      },
      "_sortVal_2": Object {
        "type": "integer",
      },
    },
    "columns": Array [
      "JobFamily",
      "Title",
      "Union",
      "Name",
      "Base",
      "TCOE",
      "Rec",
      "_depth",
      "_pivot",
      "_isRoot",
      "_sortVal_0",
      "_sortVal_1",
      "_sortVal_2",
      "_path0",
      "_path1",
    ],
  },
}
`;

exports[`initial aggtree Test 2`] = `
TableRep {
  "rowData": Array [
    Object {
      "Base": 312461,
      "JobFamily": "Executive Management",
      "Name": "Crunican, Grace",
      "Rec": 1,
      "TCOE": 399921,
      "Title": "General Manager",
      "Union": "Non-Represented",
      "_depth": 3,
      "_isRoot": 0,
      "_path0": "Executive Management",
      "_path1": "General Manager",
      "_pivot": "Crunican, Grace",
      "_sortVal_0": 1,
      "_sortVal_1": 1,
      "_sortVal_2": 1,
    },
    Object {
      "Base": 298700,
      "JobFamily": "Executive Management",
      "Name": "Dugger, Dorothy",
      "Rec": 1,
      "TCOE": 419661,
      "Title": "General Manager",
      "Union": "Non-Represented",
      "_depth": 3,
      "_isRoot": 0,
      "_path0": "Executive Management",
      "_path1": "General Manager",
      "_pivot": "Dugger, Dorothy",
      "_sortVal_0": 1,
      "_sortVal_1": 1,
      "_sortVal_2": 1,
    },
  ],
  "schema": Schema {
    "_sortedColumns": null,
    "columnIndices": Object {
      "Base": 4,
      "JobFamily": 0,
      "Name": 3,
      "Rec": 6,
      "TCOE": 5,
      "Title": 1,
      "Union": 2,
      "_depth": 7,
      "_isRoot": 9,
      "_path0": 13,
      "_path1": 14,
      "_pivot": 8,
      "_sortVal_0": 10,
      "_sortVal_1": 11,
      "_sortVal_2": 12,
    },
    "columnMetadata": Object {
      "Base": Object {
        "displayName": "Base",
        "type": "integer",
      },
      "JobFamily": Object {
        "displayName": "JobFamily",
        "type": "string",
      },
      "Name": Object {
        "displayName": "Name",
        "type": "string",
      },
      "Rec": Object {
        "type": "integer",
      },
      "TCOE": Object {
        "displayName": "TCOE",
        "type": "integer",
      },
      "Title": Object {
        "displayName": "Title",
        "type": "string",
      },
      "Union": Object {
        "displayName": "Union",
        "type": "string",
      },
      "_depth": Object {
        "type": "integer",
      },
      "_isRoot": Object {
        "type": "boolean",
      },
      "_path0": Object {
        "type": "text",
      },
      "_path1": Object {
        "type": "text",
      },
      "_pivot": Object {
        "type": "text",
      },
      "_sortVal_0": Object {
        "type": "integer",
      },
      "_sortVal_1": Object {
        "type": "integer",
      },
      "_sortVal_2": Object {
        "type": "integer",
      },
    },
    "columns": Array [
      "JobFamily",
      "Title",
      "Union",
      "Name",
      "Base",
      "TCOE",
      "Rec",
      "_depth",
      "_pivot",
      "_isRoot",
      "_sortVal_0",
      "_sortVal_1",
      "_sortVal_2",
      "_path0",
      "_path1",
    ],
  },
}
`;

exports[`t2 - basic bart table query 1`] = `
TableRep {
  "rowData": Array [
    Object {
      "Base": 75701,
      "JobFamily": "Police",
      "Name": "Ornellas, Robert",
      "TCOE": 127660,
      "Title": "Revenue Protection Guard",
      "Union": "BPOA",
    },
    Object {
      "Base": 152125,
      "JobFamily": "Safety",
      "Name": "Hardy, Leonard",
      "TCOE": 211941,
      "Title": "Chief Safety Officer",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 238269,
      "JobFamily": "Legal & Paralegal",
      "Name": "Burrows, Matthew",
      "TCOE": 321543,
      "Title": "General Counsel",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 165470,
      "JobFamily": "Legal & Paralegal",
      "Name": "Lee, Thomas",
      "TCOE": 224345,
      "Title": "Senior Attorney",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 160289,
      "JobFamily": "Legal & Paralegal",
      "Name": "Nuetzel, Victoria",
      "TCOE": 220683,
      "Title": "Senior Attorney",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 312461,
      "JobFamily": "Executive Management",
      "Name": "Crunican, Grace",
      "TCOE": 399921,
      "Title": "General Manager",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 298700,
      "JobFamily": "Executive Management",
      "Name": "Dugger, Dorothy",
      "TCOE": 419661,
      "Title": "General Manager",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 212647,
      "JobFamily": "Executive Management",
      "Name": "Oversier, Paul",
      "TCOE": 288160,
      "Title": "Assistant General Manager Operations",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 206914,
      "JobFamily": "Executive Management",
      "Name": "DeVaughn, Marcia",
      "TCOE": 281548,
      "Title": "Deputy General Manager",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 208744,
      "JobFamily": "Finance & Accounting",
      "Name": "Schroeder, Scott",
      "TCOE": 287139,
      "Title": "Controller-Treasurer",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 170383,
      "JobFamily": "Transportation Operations",
      "Name": "Crespo, Rodolfo",
      "TCOE": 238211,
      "Title": "Chief Transportation Officer",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 61184,
      "JobFamily": "Transportation Operations",
      "Name": "Smith, Westley",
      "TCOE": 118799,
      "Title": "Train Operator",
      "Union": "ATU",
    },
    Object {
      "Base": 60790,
      "JobFamily": "Transportation Operations",
      "Name": "Serrianni, Cheryl",
      "TCOE": 132030,
      "Title": "Station Agent",
      "Union": "ATU",
    },
    Object {
      "Base": 60781,
      "JobFamily": "Transportation Operations",
      "Name": "Nagel, Kevin",
      "TCOE": 105854,
      "Title": "Station Agent",
      "Union": "ATU",
    },
    Object {
      "Base": 60771,
      "JobFamily": "Transportation Operations",
      "Name": "Riggs, Joshua",
      "TCOE": 115104,
      "Title": "Station Agent",
      "Union": "ATU",
    },
    Object {
      "Base": 130035,
      "JobFamily": "Public Affairs & Marketing",
      "Name": "Hamill, Kerry",
      "TCOE": 181955,
      "Title": "Department Manager Gov't & Comm Rel",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 157300,
      "JobFamily": "Engineering & Systems Engineering",
      "Name": "Ruffa, Frank",
      "TCOE": 209709,
      "Title": "Group Manager, Capital Projects",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 152434,
      "JobFamily": "Engineering & Systems Engineering",
      "Name": "Dupont, Jean-Luc",
      "TCOE": 212535,
      "Title": "Group Manager Systems Capital Prog",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 77446,
      "JobFamily": "Maintenance, Vehicle & Facilities",
      "Name": "Hui, Sam",
      "TCOE": 128332,
      "Title": "Wayside Inspector",
      "Union": "SEIU",
    },
    Object {
      "Base": 77446,
      "JobFamily": "Maintenance, Vehicle & Facilities",
      "Name": "Reinhardt, Michael",
      "TCOE": 124962,
      "Title": "Structures Inspector",
      "Union": "SEIU",
    },
    Object {
      "Base": 77446,
      "JobFamily": "Maintenance, Vehicle & Facilities",
      "Name": "Estrada, Phillip",
      "TCOE": 123432,
      "Title": "Vehicle Inspector",
      "Union": "SEIU",
    },
    Object {
      "Base": 74068,
      "JobFamily": "Maintenance, Vehicle & Facilities",
      "Name": "Smith, Daniel",
      "TCOE": 104260,
      "Title": "Transit Vehicle Mechanic",
      "Union": "SEIU",
    },
    Object {
      "Base": 59817,
      "JobFamily": "Maintenance, Vehicle & Facilities",
      "Name": "Knight, Joshua",
      "TCOE": 113775,
      "Title": "Electrician",
      "Union": "SEIU",
    },
  ],
  "schema": Schema {
    "_sortedColumns": null,
    "columnIndices": Object {
      "Base": 2,
      "JobFamily": 4,
      "Name": 0,
      "TCOE": 3,
      "Title": 1,
      "Union": 5,
    },
    "columnMetadata": Object {
      "Base": Object {
        "displayName": "Base",
        "type": "integer",
      },
      "JobFamily": Object {
        "displayName": "JobFamily",
        "type": "string",
      },
      "Name": Object {
        "displayName": "Name",
        "type": "string",
      },
      "TCOE": Object {
        "displayName": "TCOE",
        "type": "integer",
      },
      "Title": Object {
        "displayName": "Title",
        "type": "string",
      },
      "Union": Object {
        "displayName": "Union",
        "type": "string",
      },
    },
    "columns": Array [
      "Name",
      "Title",
      "Base",
      "TCOE",
      "JobFamily",
      "Union",
    ],
  },
}
`;
